core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java

```diff
@@ -0,0 +1,212 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.tasks;
+
+import hudson.AbortException;
+import hudson.EnvVars;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.Action;
+import hudson.model.BuildListener;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.tasks.BuildWrapper;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * A generalization of {@link BuildWrapper} that, like {@link SimpleBuildStep}, may be called at various points within a build.
+ * <p>Such a build wrapper would typically be written according to make few assumptions about how it is being used.
+ * @since TODO
+ */
+@SuppressWarnings("rawtypes") // inherited
+public abstract class SimpleBuildWrapper extends BuildWrapper {
+
+    /**
+     * Called when a segment of a build is started that is to be enhanced with this wrapper.
+     * @param context a way of collecting modifications to the environment for nested steps
+     * @param build a build being run
+     * @param workspace a workspace of the build
+     * @param launcher a way to start commands
+     * @param listener a way to report progress
+     * @param initialEnvironment the environment variables set at the outset
+     * @throws IOException if something fails; {@link AbortException} for user errors
+     * @throws InterruptedException if setup is interrupted
+     */
+    public abstract void setUp(Context context, Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener, EnvVars initialEnvironment) throws IOException, InterruptedException;
+
+    /**
+     * Parameter passed to {@link #setUp} to allow an implementation to specify its behavior after the initial setup.
+     */
+    public static final class Context {
+        private Disposer disposer;
+        private final Map<String,String> env = new HashMap<String,String>();
```

Comment:
That is actually what I had originally, but it seems that `Map<String,String>` works better here. See `EnvironmentWrapper.buildEnvVars`, calling `EnvVars.putAll`, which takes a `Map`. More to the point, see https://github.com/jenkinsci/workflow-plugin/pull/37/files#diff-08a6c4ff385f29bd9c89e3395348094cR81 where `overrideAll` is called, which also takes a `Map` and indeed treats keys specially if using `+` syntax.

The one thing that would sway me is if `EnvVars` were able to carry its own set of sensitive keys, as mentioned below in the comment in `makeSensitiveBuildVariables`. This seems like a more complicated change since it would involve deprecating some things in `AbstractBuild`.